ipn/store/kubestore/store_kube.go

```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
 
 // WriteState implements the StateStore interface.
 func (s *Store) WriteState(id ipn.StateKey, bs []byte) (err error) {
-	return s.updateStateSecret(map[string][]byte{string(id): bs})
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(id)), bs)
+		}
+	}()
+	return s.updateSecret(map[string][]byte{string(id): bs}, s.secretName)
 }
 
-// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields of a Tailscale Kubernetes node's state
-// Secret.
-func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) error {
-	return s.updateStateSecret(map[string][]byte{domain + ".crt": cert, domain + ".key": key})
+// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields
+// of a Tailscale Kubernetes node's state Secret.
+func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) (err error) {
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".crt")), cert)
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".key")), key)
+		}
+	}()
+	secretName := s.secretName
+	data := map[string][]byte{domain + ".crt": cert, domain + ".key": key}
+	// If we run in cert share mode, cert and key for a DNS name are written
+	// to a separate Secret.
+	if envknob.IsCertShareReadWriteMode() {
+		secretName = sanitizeKey(domain)
+		data = map[string][]byte{keyTLSCert: cert, keyTLSKey: key}
+	}
+	return s.updateSecret(data, secretName)
 }
 
-func (s *Store) updateStateSecret(data map[string][]byte) (err error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
-	defer func() {
+// ReadTLSCertAndKey reads a TLS cert and key from memory or from a
+// domain-specific Secret. It first checks the in-memory store, if not found in
+// memory and running cert store in read-only mode, looks up a Secret.
+func (s *Store) ReadTLSCertAndKey(domain string) (cert, key []byte, err error) {
+	// Try memory first - use sanitized keys
+	certKey := sanitizeKey(domain + ".crt")
```

Comment:
Given that a Kubernetes Secret name must be a valid DNS name, can we instead error if the domain needs sanitising? I think it would make the cert state keys a bit easier to reason about knowing for sure that we don't need to worry about overlaps due to mutations.